mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,126 @@
+            Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_wrapper: &mut ImdsClientGetResourceCallback = Box::leak(callback_wrapper);
+        // callback_wrapper is safe so the raw pointer referring to that is safe as well.
+        let callback_raw_ptr: *mut libc::c_void =
+            callback_wrapper as *mut ImdsClientGetResourceCallback as *mut libc::c_void;
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
+
+        // Safety: move the reference where the data store in memory out of box potential leaks memory,
+        // however, the binding callback function will drop it once the corresponding crt function finished the execution.
+        // See `extern "C" fn imds_client_get_resource_callback`.
+        unsafe {
+            let result =
+                aws_imds_client_get_instance_type(self.inner.as_ptr(), get_resource_callback_fn_ptr, callback_raw_ptr);
+            Ok(result)
+        }
+    }
+}
+
+impl Drop for ImdsClient {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` points to a valid aws_imds_client, and when this [ImdsClient] is dropped.
+        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
+    }
+}
+
+impl ImdsClientConfig {
+    /// Create a new [ImdsClientConfig] with default options.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Client bootstrap used for common staples such as event loop group, host resolver, etc.
+    pub fn client_bootstrap(&mut self, client_bootstrap: ClientBootstrap) -> &mut Self {
+        self.inner.bootstrap = client_bootstrap.inner.as_ptr();
+        self.client_bootstrap = Some(client_bootstrap);
+        self
+    }
+
+    /// Retry strategy used to reschedule failed requests
+    pub fn retry_strategy(&mut self, retry_strategy: RetryStrategy) -> &mut Self {
+        self.inner.retry_strategy = retry_strategy.inner.as_ptr();
+        self.retry_strategy = Some(retry_strategy);
+        self
+    }
+}
+
+/// Callback trait for when finished getting ec2 instance type.
+type OnGetResource = Box<dyn FnOnce(Result<OsString, Error>)>;
+struct ImdsClientGetResourceCallback(OnGetResource);
+
+/// Rust binding for CRT's callback function `aws_imds_client_on_get_resource_callback_fn`.
+unsafe extern "C" fn imds_client_get_resource_callback(
+    resource: *const aws_byte_buf,
+    error_code: i32,
+    user_data: *mut libc::c_void,
+) {
+    let resource_slice = std::slice::from_raw_parts((*resource).buffer, (*resource).len);
+    let resource = OsStr::from_bytes(resource_slice).to_os_string();
+    let callback = Box::from_raw(user_data as *mut ImdsClientGetResourceCallback).0;
+    let result = if 0 != error_code {
+        Err(Error::from(error_code))
+    } else {
+        Ok(resource)
+    };
+
+    callback(result);
+}
```

Comment:
Maybe we can improve the flow here a bit. Some stuff we don't need to do if we encounter the error case.

```suggestion
/// Rust binding for CRT's callback function `aws_imds_client_on_get_resource_callback_fn`.
unsafe extern "C" fn imds_client_get_resource_callback(
    resource: *const aws_byte_buf,
    error_code: i32,
    user_data: *mut libc::c_void,
) {
    let callback = Box::from_raw(user_data as *mut ImdsClientGetResourceCallback).0;
    
    let result = if 0 != error_code {
        Err(Error::from(error_code))
    } else {
        let resource_slice = std::slice::from_raw_parts((*resource).buffer, (*resource).len);
        let resource = OsStr::from_bytes(resource_slice).to_os_string();
        Ok(resource)
    };
    
    callback(result)
}
```

Reply:
will fix